contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,439 @@
+the whole request.
+
+For watch operations, as described in requirements, we also need to periodically
+(or maybe lazily) verify whether their ACLs didn't change (and user is still
+allowed to watch requested objects). However, as also mentioned above, this
+periodic checking isn't specific to bulk operations and we need to support the
+same mechanism for regular watch too. We will just ensure that this mechanism
+tracking and periodically verifying ACLs is implemented low enough in apiserver
+machinery so that we will be able to reuse exactly the same one for the purpose
+of bulk watch operations.
+For watch request, we will support partial rejections. The exact details of it
+will be described together with dynamic watch description below.
+
+
+### Dynamic watch
+
+ As mentioned in the Proposal section, we will implement bulk watch that will
+allow for dynamic subscribtion/unsubscribtion for (sets of) objects on top of
+websockets protocol.
+
+ Note that we already support websockets in the regular Kubernetes API for
+watch requests (in addition to regular http requests), so for the purpose of
+bulk watch we will be extending websocket support.
+
+ The the high level, the propocol will look:
+1. client opens a new websocket connection to a bulk watch endpoint to the
+server via ghttp GET
+1. this results in creating a single channel that is used only to handle
+communication for subscribing/unsubscribing for watches; no watch events are
+delivered via this particular channel.  
+*TODO: Consider switching to two channels, one for incoming and one for
+outgoing communication*
+1. to subscribe for a watch of a given (set of) objects, user sends `Watch`
+object over the channel; in response a new channel is created and the message
+with the channel identifier is send back to the user (we will be using integers
+as channel identifiers).
+1. once subscribed, all objects matching a given selector will be send over
+the newly created channel
+1. to stop watching for a given (set of) objects, user sends `CloseWatch`
+object over the the channel; in response the corresponding watch is broken and
+corresponding channel within websocket is closed
+1. once done, user can close the whole websocket connection (this results in
+breaking all still opened channels and corresponding watches).
+
+ With that high level protocol, there are still multiple details that needs
+to be figured out. First, we need to define `Watch` and `CloseWatch`
+message. We will solve it with the single `Request` object:
+```
+type Request struct {
+	// Only one of those is set.
+	Watch      Watch
+	CloseWatch CloseWatch
+}
+
+type Identifier int64
+
+// Watch request for objects matching a given selector.
+type Watch struct {
+	Selector GetOperation
+}
+
+// Request to stop watching objects from a watch identified by the channel.
+type CloseWatch struct {
+	Channel Identifier
+}
+
+// Depending on the request, channel that was created or deleted.
+type Response struct {
```

Comment:
My preference is to have an int that the client increments with every request which the server echos back in the response.